### DISCUSSION TITLE
Unable to grep output of command

### DISCUSSION THREAD

**Discussion by therealkris** [community/community#25935] [Actions]
<p>I have a step that runs a script to deploy a revision to google cloud endpoints and then get the configuration ID from that. The script works perfectly in the local environment but fails to grab the configuration ID in the github actions runner. The relevant lines in the script follow:</p>
<pre><code class="lang-auto"># Deploy updated endpoints configuration
gcloud endpoints services deploy openapi.json --project ${PROJECT}

# Get latest configuration
services=$(gcloud endpoints configs list --service=${SERVICE} --limit 1)
# Extract the configuration ID from the services list - The format is YYYY-MM-DDrXX
config_id=$(echo ${services} | grep -ioE '\d+\-\d+\-\d+[a-z]\d+')
echo "ConfigID: ${config_id}"
</code></pre>
<p>On the last line, the <code>config_id</code> should look similar to 2020-07-21r21. This works as expected in the local environment, but not in the action runner. If I echo <code>services</code>, I see <code>CONFIG_ID SERVICE_NAME 2020-07-21r20 ***</code> (the *** is the service being passed in as a command to the script via a secret env variable)</p>
<p>The step is:</p>
<pre><code class="lang-auto">- name: Get Config ID
  run: ./config_id -p ${{ secrets.PROJECT }} -s ${{ secrets.SERVICE }} -n ${{ secrets.NAME }}
</code></pre>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/therealkris">@therealkris</a>,</p>
<p>In GitHub Actions workflow, <a href="https://docs.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables#about-environment-variables" rel="nofollow noopener">environment variables are case-sensitive</a>.</p>
<p>From your comments,</p>
<blockquote>
<p><strong>If I echo services, I see CONFIG_ID SERVICE_NAME 2020-07-21r20 ***</strong></p>
</blockquote>
<p>I notice the variable name “<strong>CONFIG_ID</strong>” is completely <strong>uppercase</strong>.</p>
<p>But in the script,</p>
<blockquote>
<p><strong>echo "ConfigID: ${config_id}"</strong></p>
</blockquote>
<p>you are using the variable name “<strong>config_id</strong>” that is completely <strong>lowercase</strong>.</p>
<p>Please change the variable name to uppercase (<strong>echo “ConfigID: ${CONFIG_ID}”</strong>) in the script to see if the problem can be solved.</p>

**Comment by therealkris** [comment]
<p><a class="mention" href="/brightran">@brightran</a> That’s the header text in the command output, not an actual variable. The variable <code>config_id</code> comes from this line: <code>config_id=$(echo ${services} | grep -ioE '\d+\-\d+\-\d+[a-z]\d+')</code>. It takes that output, finds the text that matches the regex (in this case <code>2020-07-21r20</code>), and sets that value as <code>config_id</code>. Remember, this script works perfectly when run locally. When run as part of the actions workflow, though, this step fails.</p>
<p>As a note, though, I updated the whole thing from a shell script to a python script and it is now working fine, so my immediate issue is resolved. The python script does not have the same issue with the regular expression matching. What I’d be interested in knowing, however, is what the issue is in the workflow run that causes the <code>grep</code> to return nothing. I suspect it has something to do with the <code>***</code> in the gcloud command output, which corresponds to a service url that is stored in a secret, but I don’t know.</p>

**Comment by BrightRan** [comment, marked as answer]
<p><a class="mention" href="/therealkris">@therealkris</a>,</p>
<blockquote>
<p>The variable  <code>config_id</code>  comes from this line:  <code>config_id=$(echo ${services} | grep -ioE '\d+\-\d+\-\d+[a-z]\d+')</code> .</p>
</blockquote>
<p>Ah, yes, I missed it.<br>
Now I seem find the reason that way the command fails to get the configuration ID.</p>
<p>In the <strong>grep</strong> command, it seems does not recognize the metacharacter ‘<strong>\d</strong>’ in the regular expression “<strong>\d+\-\d+\-\d+[a-z]\d+</strong>”.<br>
I know, generally many people (include me) are used to using the metacharacter ‘<strong>\d</strong>’ instead of '<strong>[0-9]</strong>‘ to represent the digital characters in the regular expression.<br>
However, I find that some command line interfaces do not recognize the metacharacter ‘<strong>\d</strong>’, and they seem to prefer the metacharacter '<strong>[0-9]</strong>‘.<br>
Maybe in some environments, the two metacharacters are supported in the commands.</p>
<p>I tested on my local environment (Windows 10, Bash), and only when I used the metacharacter ‘<strong>[0-9]</strong>’, the configuration ID can be fetched. When using '<strong>\d</strong>‘, it returned nothing.</p>
<ul>
<li>
<strong>[Test 1]</strong><br>
script:</li>
</ul>
<pre><code class="lang-bash">services='CONFIG_ID SERVICE_NAME 2020-07-21r20 this the service name'
# using the metacharacter '[0-9]'
config_id=$(echo ${services} | grep -ioE '[0-9]+\-[0-9]+\-[0-9]+[a-z][0-9]+')
echo "config_id = ${config_id}"
</code></pre>
<p>result:<br>
<img src="https://user-images.githubusercontent.com/46307996/181094197-077bc3f3-faca-4573-ae11-b7aecd0e3351.png" alt="0-9" data-base62-sha1="43a5uq0BDmsSZ1aazG5hekoKedB" width="560" height="190"></p>
<ul>
<li>
<strong>[Test 2]</strong><br>
script:</li>
</ul>
<pre><code class="lang-bash">services='CONFIG_ID SERVICE_NAME 2020-07-21r20 this the service name'
# using the metacharacter '\d'
config_id=$(echo ${services} | grep -ioE '\d+\-\d+\-\d+[a-z]\d+')
echo "config_id = ${config_id}"
</code></pre>
<p>result:<br>
<img src="https://user-images.githubusercontent.com/46307996/181094203-a4c4d35d-83a0-473a-8227-c3c08dc2df36.png" alt="-d" data-base62-sha1="aYNXbqyxNe6zTGgeyEYuK8SMFsk" width="501" height="189"></p>

**Comment by therealkris** [comment]
<p>Ah that’s interesting. I didn’t consider that the regex would be supported differently across environments. After some testing, I’ve found that on my that macOS command line, <code>\d</code> works, but in the ubuntu command line, it does not and only <code>[0-9]</code> works.</p>
<p>Thanks for the tip!</p>

**Comment by airtower-luna** [comment]
<p>The issue is the use of <code>-E</code>, which means to use Unix extended regular expressions. Use <code>-P</code> instead for Perl-compatible regular expressions, and <code>\d</code> works. Or alternatively use <code>[[:digit:]]</code> for the equivalent character class in extended regular expressions. :wink:</p>
<p>I guess grep on MacOS has different defaults there.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/airtower-luna">@airtower-luna</a>,</p>
<p>Thanks for your reminder.<br>
Yes, when using the option ‘<strong>-P</strong>’ (<strong>–perl-regexp</strong>), both the ‘<strong>\d</strong>’ and ‘<strong>[0-9]</strong>’ can work as expected.</p>
<p><a class="mention" href="/therealkris">@therealkris</a>,</p>
<p>I just remembered that there are many regular expression flavors. The popular regular expression flavors include <a href="https://www.regular-expressions.info/perl.html" rel="nofollow noopener"><strong>Perl</strong></a>, <a href="https://www.regular-expressions.info/pcre.html" rel="nofollow noopener"><strong>PCRE</strong></a>, <a href="https://www.regular-expressions.info/php.html" rel="nofollow noopener"><strong>PHP</strong></a>, <a href="https://www.regular-expressions.info/dotnet.html" rel="nofollow noopener"><strong>.NET</strong></a>, <a href="https://www.regular-expressions.info/java.html" rel="nofollow noopener"><strong>Java</strong></a>, <a href="https://www.regular-expressions.info/javascript.html" rel="nofollow noopener"><strong>JavaScript</strong></a>, <a href="https://www.regular-expressions.info/xregexp.html" rel="nofollow noopener"><strong>XRegExp</strong></a>, <a href="https://www.regular-expressions.info/vbscript.html" rel="nofollow noopener"><strong>VBScript</strong></a>, <a href="https://www.regular-expressions.info/python.html" rel="nofollow noopener"><strong>Python</strong></a>, <a href="https://www.regular-expressions.info/ruby.html" rel="nofollow noopener"><strong>Ruby</strong></a>, <a href="https://www.regular-expressions.info/delphi.html" rel="nofollow noopener"><strong>Delphi</strong></a>, <a href="https://www.regular-expressions.info/rlanguage.html" rel="nofollow noopener"><strong>R</strong></a>, <a href="https://www.regular-expressions.info/tcl.html" rel="nofollow noopener"><strong>Tcl</strong></a>, <a href="https://www.regular-expressions.info/posix.html" rel="nofollow noopener"><strong>POSIX</strong></a>, etc…<br>
Each regular expression flavor is the syntax and behavior of a particular engine.</p>
<p>When using the option ‘<strong>-E</strong>’ (<strong>–extended-regexp</strong>), it is specified by <strong>POSIX</strong>.<br>
<a href="https://www.regular-expressions.info/posixbrackets.html" rel="nofollow noopener">One key syntactic difference is that the <strong>backslash</strong> (<strong>\</strong>) is NOT a metacharacter in a POSIX bracket expression. So in POSIX, the regular expression ‘<strong>[\d]</strong>’ matches a ‘<strong>\</strong>’ or a ‘<strong>d</strong>’.</a></p>

**Comment by therealkris** [comment]
<p><a class="mention" href="/airtower-luna">@airtower-luna</a> <a class="mention" href="/brightran">@brightran</a> Yep, I see that now. That you guys for the help! I’m all set.</p>
